cmd/kube-apiserver/app/aggregator.go

```diff
@@ -304,13 +309,78 @@ func apiServicesToRegister(delegateAPIServer genericapiserver.DelegationTarget,
 			continue
 		}
 
+		hash, err := getGroupVersionHash(curr, delegateAPIServer.UnprotectedHandler())
+		if err != nil {
+			klog.Warning(err)
+		}
 		apiService := makeAPIService(schema.GroupVersion{Group: tokens[2], Version: tokens[3]})
 		if apiService == nil {
 			continue
 		}
-		registration.AddAPIServiceToSyncOnStart(apiService)
+		registration.AddAPIServiceToSyncOnStart(apiService, hash)
 		apiServices = append(apiServices, apiService)
 	}
 
 	return apiServices
 }
+
+func getGroupVersionHash(path string, handler http.Handler) (string, error) {
+	// TODO: This can be optimized by using HEAD since we are only interested in the ETag
+	req, err := http.NewRequest("GET", path, nil)
+	if err != nil {
+		return "", err
+	}
+	writer := newInMemoryResponseWriter()
+	handler.ServeHTTP(writer, req)
+	if writer.respCode != http.StatusOK {
+		return "", fmt.Errorf("Error requesting has for group-version under %s", path)
+	}
+	etag := writer.Header().Get("Etag")
+	etag, err = strconv.Unquote(etag)
+	if err != nil {
+		return "", err
+	}
+	return etag, nil
+}
+
+// inMemoryResponseWriter is a http.Writer that keep the response in memory.
+type inMemoryResponseWriter struct {
+	writeHeaderCalled bool
+	header            http.Header
+	respCode          int
+	data              []byte
+}
+
+// This is duplicated from the OpenAPI downloader
+// TODO: Maybe move this into a utility library
```

Comment:
Are there plans of moving this to a common location accessible to all of the open api downloader, this module, and the crdregistration_controller (from pkg/controlplane)? Does something like this already exist?